### PR TITLE
Add day14

### PR DESCRIPTION
**Pull request by mtsochacki**
Hopefully this one is not using any "for i loops" that it doesn't have to use...

### CODE REVIEW

File: 2021/day14/day14.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.io.File;
+
+public class day14 {
+    public static class Rules {
+        String pair;
+        String single;
+    }
```
Comment (Lammatian) [inline 2021/day14/day14.java:11]: I would call it a `Rule`, it's not like this represents multiple of them. Also, to make this more readable, I would rename `single` to `outcome` or `result` or something along these lines.

File: 2021/day14/day14.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        } finally {
+            sc.close();
+        }
+        return input;
+    }
+
+    public static HashMap<String, Long> splitToChunks(String template) {
+        HashMap<String, Long> chunks = new HashMap<>();
+        for (int i = 0; i < template.length() - 1; i++) {
+            chunks.merge(String.valueOf(template.charAt(i))
```
Comment (Lammatian) [inline 2021/day14/day14.java:50]: `template.substring(i, i + 2)` instead of these `valueOf(charAt)`s?

File: 2021/day14/day14.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        HashMap<String, Long> letters = countLetters(template);
+        for (int i = 0; i < steps; i++) {
+            HashMap<String, Long> newChunks = new HashMap<>();
+            for (Rules rule : input) {
+                if (chunks.containsKey(rule.pair)) {
+                    letters.merge(rule.single, chunks.get(rule.pair), Long::sum);
+                    newChunks.merge(String.valueOf(rule.pair.charAt(0))
+                            + rule.single, chunks.get(rule.pair), Long::sum);
+                    newChunks.merge(String.valueOf(rule.single
+                            + rule.pair.charAt(1)), chunks.get(rule.pair), Long::sum);
```
Comment (Lammatian) [inline 2021/day14/day14.java:77]: In my order of preference, I would either create the new pairs in lines above the `merge`s, i.e.

```java
leftPair = rule.pair.substring(0, 1) + rule.single;
newChunks.merge(leftPair, chunks.get(rule.pair), Long::sum);
// same for right pair
```

or do the splitting differently:
```java
newChunks.merge(
    rule.pair.substring(0, 1) + rule.single,
    chunks.get(rule.pair),
    Long::sum);
// same for right pair
```

**Review by Lammatian — CHANGES_REQUESTED** [review]
Minor changes required here, well done!

File: 2021/day14/day14.java (hunk 2 above)
Comment (mtsochacki) [reply 2021/day14/day14.java:50]: I did not know this method exists :+1:

File: 2021/day14/day14.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.io.File;
+
+public class day14 {
+    public static class Rule {
+        String pair;
+        String outcome;
+    }
```
Comment (Lammatian) [inline 2021/day14/day14.java:11]: Add a space after the end of the statement. Same goes for lines 25, 45 and 53

File: 2021/day14/day14.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    // Read polymer template e.g. NCNBCHB
```
Comment (Lammatian) [inline 2021/day14/day14.java:12]: This is not descriptive - the name of the method already mentions all of this. I would say this method does not require a description

File: 2021/day14/day14.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            sc = new Scanner(new File(filename));
+            template = sc.next();
+        } catch (Exception e) {
+            System.out.println("Something went wrong" + e);
+        } finally {
+            sc.close();
+        }
+        return template;
+    }
+    // Read list of rules e.g. CH -> B
```
Comment (Lammatian) [inline 2021/day14/day14.java:26]: This method could use a description as there is some kind of transformation happening, but the current comment does not capture this. I would say something like

```java
/*
 * Given the name of a file containing a list of rules for creating
 * new elements in the form of arrow-delimited strings, e.g. "CH -> B",
 * read and transform them into a list of `Rule`s
 */
```

File: 2021/day14/day14.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    public static ArrayList<Rule> readRules(String filename) {
+        Scanner sc = null;
+        ArrayList<Rule> input = new ArrayList<>();
+        try {
+            sc = new Scanner(new File(filename));
+            sc.useDelimiter("\\n| -> ");
+            while (sc.hasNext()) {
+                Rule line = new Rule();
```
Comment (Lammatian) [inline 2021/day14/day14.java:34]: I would rename this from `line` into e.g. `rule`. `line` is misleading

File: 2021/day14/day14.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                input.add(line);
+            }
+        } catch (Exception e) {
+            System.out.println("Something went wrong" + e);
+        } finally {
+            sc.close();
+        }
+        return input;
+    }
+    // Split template into two-letter chunks NCNBCHB -> NC, CN, NB, BC, CH...
```
Comment (Lammatian) [inline 2021/day14/day14.java:46]: I would move the example to a new line:

```java
/*
 * Count the occurrences of overlapping two-letter chunks in the template.
 * 
 * E.g. given a template
 * NCCNC
 * Transform it into a map
 * {"NC": 2, "CN": 1, "CC": 1}
 */

**Review by Lammatian — CHANGES_REQUESTED** [review]
I like the idea of documenting your code. Although remember that the more you can describe with the code itself, the better.

File: 2021/day14/day14.java (hunk 6 above)
Comment (mtsochacki) [reply 2021/day14/day14.java:26]: Ok, I am just going to use that. There is no way I am going to come up with a better description

File: 2021/day14/day14.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+        /*
+         * Given the name of a file containing a list of rules for creating
+         * new elements in the form of arrow-delimited strings, e.g. "CH -> B",
+         * read and transform them into a list of `Rule`s
+         */
+    public static ArrayList<Rule> readRules(String filename) {
+        Scanner sc = null;
+        ArrayList<Rule> input = new ArrayList<>();
```
Comment (Lammatian) [inline 2021/day14/day14.java:34]: This could also be renamed from `input` to `rules`

File: 2021/day14/day14.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+    /*
+     * Split the template into individual letters and count the occurrences
+     * of them in the template.
+     *
+     * E.g. given a template
+     * NCNBCHB
+     * Transform it into a map
+     * {"N":2, "C":2, "B":2, "H":1}
+     */
```
Comment (Lammatian) [inline 2021/day14/day14.java:77]: Try to keep it in the bounds of 80 characters per line from now on. This is especially easy in the documentation where you don't have to care about how it looks. In VSCode you can probably even turn on auto wrap on column 80 or something.

**Review by Lammatian — APPROVED** [review]
This is mergeable, good job :)

File: 2021/day14/day14.java (hunk 10 above)
Comment (mtsochacki) [reply 2021/day14/day14.java:77]: Screenshot from 2021-12-20 18-26-45
It is within 80 on my screen with the exception of 95-104, but I will have a look at auto wrap and keep it in mind in the future.

File: 2021/day14/day14.java (hunk 9 above)
Comment (mtsochacki) [reply 2021/day14/day14.java:34]: I thought about it too, I am going to rename it

**Pull request merged**